src/ocrd/processor/base.py

```diff
@@ -188,20 +193,122 @@ def verify(self):
         """
         return True
 
+    def setup(self) -> None:
+        """
+        Prepare the processor for actual data processing,
+        prior to changing to the workspace directory but
+        after parsing parameters.
+
+        (Override this to load models into memory etc.)
+        """
+        pass
+
+    @deprecated(version='3.0', reason='process() should be replaced with process_page() and process_workspace()')
     def process(self) -> None:
         """
-        Process the :py:attr:`workspace` 
+        Process all files of the :py:attr:`workspace` 
         from the given :py:attr:`input_file_grp`
         to the given :py:attr:`output_file_grp`
-        for the given :py:attr:`page_id`
+        for the given :py:attr:`page_id` (or all pages)
         under the given :py:attr:`parameter`.
         
         (This contains the main functionality and needs to be overridden by subclasses.)
         """
         raise NotImplementedError()
 
+    def process_workspace(self, workspace: Workspace) -> None:
+        """
+        Process all files of the given ``workspace``,
+        from the given :py:attr:`input_file_grp`
+        to the given :py:attr:`output_file_grp`
+        for the given :py:attr:`page_id` (or all pages)
+        under the given :py:attr:`parameter`.
+
+        (This will iterate over pages and files, calling
+        :py:meth:`process_page`, handling exceptions.)
+        """
+        # assert self.input_file_grp is not None
+        # assert self.output_file_grp is not None
+        # input_file_grps = self.input_file_grp.split(',')
+        # for input_file_grp in input_file_grps:
+        #     assert input_file_grp in workspace.mets.file_groups
+        log = getLogger('ocrd.processor.base')
+        with pushd_popd(workspace.directory):
+            self.workspace = workspace
+            try:
+                # FIXME: add page parallelization by running multiprocessing.Pool (#322)
```

Comment:
I hope the Pool workers would be also controllable by an environment variable to ease resource distribution in HPC environments.

Reply:
That's the idea, exactly.